### PR TITLE
[cmake] use -Werror for double-promotion, missing-field-initializers, and sign-compare

### PR DESCRIPTION
**Pull request by lrusak**
We seem to continue adding warnings and continue having to fix them. Let's make these `-Werror` now to avoid adding new warnings.

I'm not sure if there is any negative repercussions at this time.

I had to disable the behaviour for a couple of the external libraries. This seems like an ok solution for now.

Note that this is only for posix platforms. I'm not sure what the windows changes would be.

### CODE REVIEW

**Comment by AlwinEsch** [comment]
I think it is good to use more warnings and it makes the code more secure, especially since gcc 11 itself already has a higher definition and is then noticed more quickly on older ones.

I'm just a little torn because of the depends, I think it's great to see warnings there because you may then recognize background problems. Stupid only that with some depends a "no-error" is necessary, hope is not lost with updates to check.

I wouldn't worry that much about Windows, because the warning declaration is very high there anyway.

**Comment by lrusak** [comment]
I went a bit of a different direction with this. We can't really use the cmake `add_options` like we do for warnings as they apply to all code built by cmake (including cmake external projects, some dependencies, and the lib/ folder).

I've added a couple comments about where options apply so when adding new ones.

The flags now look like this for me. Note that later options override earlier ones.
```
CXX_FLAGS = -Wall -Wdouble-promotion -Wmissing-field-initializers -Wsign-compare -g -D_DEBUG <snip> -Werror=double-promotion -Werror=missing-field-initializers -Werror=sign-compare -std=c++14
```

**Comment by lrusak** [comment]
jenkins build this please

**Comment by wsnipex** [comment]
looks sane to me, only question is if we should make CORE_COMPILE_OPTIONS an actual option or user overridable.
Did you test, if setting CXXFLAGS or CMAKE_CXX_FLAGS allows turning those off?

**Review by garbear — APPROVED** [review]

**Comment by lrusak** [comment]
> looks sane to me, only question is if we should make CORE_COMPILE_OPTIONS an actual option or user overridable. Did you test, if setting CXXFLAGS or CMAKE_CXX_FLAGS allows turning those off?

That's a good point. I'm not sure what the best way to handle that is. It seems that using `CXXFLAGS` env variable or `-DCMAKE_CXX_FLAGS` configure variable both prepend the flags. That won't work as latter options override prior options.

So that maybe makes it an options that can be overridden (although I don't like adding new options). Do we just leave it as is for now and see if it becomes problematic?

File: cmake/scripts/common/ArchSetup.cmake [hunk 1]
```diff
@@ -148,8 +148,24 @@ if(NOT DEFINED NEON OR NEON)
...
+    -Wmissing-field-initializers
+    -Wsign-compare
+  )
+  add_options(ALL_LANGUAGES DEBUG
+    -g
+    -D_DEBUG
+  )
+
+  # these options affect only core code
+  set(CORE_COMPILE_OPTIONS
```
Comment (wsnipex) [inline cmake/scripts/common/ArchSetup.cmake:164]: let's allow the user to override:
```suggestion
  if(NOT CORE_COMPILE_OPTIONS)
    set(CORE_COMPILE_OPTIONS
```

should be sufficient, without adding an explicit option

Comment (lrusak) [reply cmake/scripts/common/ArchSetup.cmake]: I'm fine with that 👍

**Review by wsnipex — APPROVED** [review]

**Comment by lrusak** [comment]
squashed relevant fixups are reorder the commits to have all the fix commits before enabling `-Werror`

**Review by AlwinEsch — APPROVED** [review]

**Comment by AlwinEsch** [comment]
jenkins build this with addons please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
Have you somewhere the link about Jenkins and binary addons, want to make sure that it not affects them?

**Comment by mkreisl** [comment]
```
grep warning.*implicit build.log | wc -l
66
```
(32bit Debian Bullseye build for Raspberry Pi4)
This commit is once again a good example of a senseless waste of time and resources here.
 I'm sorry, but I had to say that.

**Comment by thexai** [comment]
> We seem to continue adding warnings and continue having to fix them.

Yeah!
https://github.com/xbmc/xbmc/commit/76105534ad94d9b25832520231f615cc4a6109fb has caused 4 **new** Visual Studio warnings:

https://jenkins.kodi.tv/view/Windows/job/WIN-64/19260/msbuild/folder.917671344/

https://jenkins.kodi.tv/view/Windows/job/WIN-64/19260/msbuild/folder.917671344/source.9082885b-d4cb-4305-a96a-91e286d7455b/#57

**Comment by lrusak** [comment]
> Have you somewhere the link about Jenkins and binary addons, want to make sure that it not affects them?

@AlwinEsch 
https://jenkins.kodi.tv/view/Automation/job/BuildMultiWithAddons-PR/lastFailedBuild/

**Comment by lrusak** [comment]
> ```
> 
> grep warning.*implicit build.log | wc -l
> 
> 66
> 
> ```
> 
> (32bit Debian Bullseye build for Raspberry Pi4)
> 
> This commit is once again a good example of a senseless waste of time and resources here.
> 
>  I'm sorry, but I had to say that.

I'm a little lost. What is the point of this comment?

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
Let's see what happens here.... 🙃

**Comment by howie-f** [comment]
_dang_
```
GUIFontTTF.cpp: In member function 'std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecText&)':
GUIFontTTF.cpp:666:49: error: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Werror=sign-compare]
  666 |       for (unsigned int j = lastSetIndex + 1; j < i; j++)
      |                                               ~~^~~
cc1plus: some warnings being treated as errors
```

**Comment by Portisch** [comment]
```
/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1173:44: error: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Werror=sign-compare]
 1173 |     (stream->disabled && stream->ExtraSize != m_pFormatContext->streams[pPacket->iStreamId]->codecpar->extradata_size))
      |                          ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Pull request merged**